src/devices/ikea_sparsnas.c

```diff
@@ -0,0 +1,251 @@
+last six digits are your sender ID. Eg "400 617 633" gives you the sender id 
+617633. This number goes into IKEA_SPARSNAS_SENSOR_ID in this file.
+
+
+The data is sent using GFSK modulation. It requires PD_MIN_PULSE_SAMPLES in 
+pulse_detect.h to be lowered to 5 to be able to demodulate at 250kS/s. The
+preamble is optimally 4 bytes of 0XAA. Then the sync word 0xD201. Here only 
+the last 2 bytes of the 0xAA preamble is checked, as the first ones seems
+to be corrupted quite often. There are plenty of integrety checks made on 
+the demodulated package which makes this compromise OK.
+
+Packet structure according to: https://github.com/strigeus/sparsnas_decoder 
+(with some changes by myself)
+
+0:  uint8_t length;        // Always 0x11
+1:  uint8_t sender_id_lo;  // Lowest byte of sender ID
+2:  uint8_t unknown;       // Not sure
+3:  uint8_t major_version; // Always 0x07 - the major version number of the sender.
+4:  uint8_t minor_version; // Always 0x0E - the minor version number of the sender.
+5:  uint32_t sender_id;    // ID of sender
+9:  uint16_t sequence;     // Sequence number of current packet
+11: uint16_t effect;       // Current effect usage
+13: uint32_t pulses;       // Total number of pulses
+17: uint8_t battery;       // Battery level, 0-100%
+18: uint16_t CRC;          // 16 bit CRC of bytes 0-17
+
+The packet's integrety can be checked with the 16b CRC at the end of the packet. 
+There are also several other ways to check the integrety of the package. 
+ - (preamble)
+ - CRC
+ - The decrypted sensor ID
+ - the constant bytes at 0, 3 and 4
+
+The decryption, CRC is calculation, value extraction and interpretation is 
+taken from https://github.com/strigeus/sparsnas_decoder and adapted to 
+this application. Many thanks to strigeus!
+
+Most other things are from https://github.com/kodarn/Sparsnas which is an
+amazing repository of the IKEA Sparsnäs. Everything is studied with greay
+detail. Many thanks to kodarn!
+
+*/
+
+
+
+#define IKEA_SPARSNAS_SENSOR_ID 617633
+#define IKEA_SPARSNAS_PULSES_PER_KWH 1000
+
+// No settings required below this point
+
+
+#include "decoder.h"
+#define IKEA_SPARSNAS_MESSAGE_BITLEN 160    // 20 bytes incl 8 bit length, 8 bit address, 128 bits data, and 16 bits of CRC. Excluding preamble and sync word
+#define IKEA_SPARSNAS_MESSAGE_BYTELEN    ((IKEA_SPARSNAS_MESSAGE_BITLEN + 7) / 8)
+
+#define IKEA_SPARSNAS_PREAMBLE_BITLEN 32
+static const uint8_t preamble_pattern[4] = {0xAA, 0xAA, 0xD2, 0x01};
+
+
+// Graciously copied from https://github.com/strigeus/sparsnas_decoder
+static uint16_t ikea_sparsnas_crc16(const uint8_t *data, size_t n) {
+  uint16_t crcReg = 0xffff;
+  size_t i, j;
+  for (j = 0; j < n; j++) {
+    uint8_t crcData = data[j];
+    for (i = 0; i < 8; i++) {
+      if (((crcReg & 0x8000) >> 8) ^ (crcData & 0x80))
+        crcReg = (crcReg << 1) ^ 0x8005;
+      else
+        crcReg = (crcReg << 1);
+      crcData <<= 1;
+    }
+  }
+  return crcReg;
+}
```

Comment:
We have our own crc functions. So eventually this needs to be replaced.

Reply:
OK. Changed to crc16 in utils.c and verified that it works.